File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int getMaxProfit(vector<int> &stocks){  
    int minBuy = 1e9, maxProfit = 0;

    for(int i=0; i<stocks.size(); i++) // Time Comp: O(n) and SC O(1)
    {
        // selling at curr day
        maxProfit = max(maxProfit, stocks[i] - minBuy);

        // update min, for future
        minBuy = min(minBuy, stocks[i]);
    }

    return maxProfit;
}

    int maxProfit(vector<int>& prices) {
        return getMaxProfit(prices);
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
      int profit = 0;
      int mxprofit = 0;
      int currVal = prices[0];
      for(int i=1; i<prices.length; i++){
          profit = prices[i] - currVal;
          mxprofit = Math.max(profit, mxprofit);
          currVal = Math.min(currVal, prices[i]);
      }  
      return mxprofit;
    }
}